File: 分类/图/题目/leetcode_210.py
# 现在你总共有 n 门课需要选，记为 0 到 n-1。
# 在选修某些课程之前需要一些先修课程。 例如，想要学习课程 0 ，你需要先完成课程 1 ，我们用一个匹配来表示他们: [0,1]
# 给定课程总量以及它们的先决条件，返回你为了学完所有课程所安排的学习顺序。
# 可能会有多个正确的顺序，你只要返回一种就可以了。如果不可能完成所有课程，返回一个空数组。

from typing import DefaultDict, List

# leetcode - 210
# 图 | 邻接表 | bfs
class Solution:
    # 时间复杂度: O(V*E)
    # 空间复杂度: O(V*E)
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        if not prerequisites: 
            return [i for i in range(0, numCourses)]

        # 我依赖谁 | 入度表
        inDeg = DefaultDict(list)
        # 谁依赖我 | 出度表
        outDeg = DefaultDict(list)

        for item in prerequisites:
            inDeg[item[0]].append(item[1])
            outDeg[item[1]].append(item[0])
        
        queue = []
        result = []

        # 搜索无依赖课程
        for item in range(numCourses):
            if len(inDeg[item]) == 0:
                queue.append(item)
        
        # 有环 - 返回空
        if len(queue) == 0:
            return []
        
        while queue:
            item = queue.pop(0)
            result.append(item)
            outs = outDeg[item]

            # 更新队列
            for out in outs:
                if len(inDeg[out]) == 1:
                    queue.append(out)
                elif len(inDeg[out]) > 1:
                    inDeg[out].remove(item)
                   
        return result if len(result) == numCourses else []
    
so = Solution()

prerequisites = [[1,0],[2,0],[3,1],[3,2]]
result = so.findOrder(4, prerequisites)
print(result) 
    

File: Leetcode(Swift)/Graph/Leetcode_210.swift
/*
 There are a total of numCourses courses you have to take, labeled from 0 to numCourses - 1.
 You are given an array prerequisites where prerequisites[i] = [ai, bi] indicates that you must take course bi first if you want to take course ai.

 For example, the pair [0, 1], indicates that to take course 0 you have to first take course 1.
 Return the ordering of courses you should take to finish all courses. If there are many valid answers,
 return any of them. If it is impossible to finish all courses, return an empty array.
 */
import Foundation

final class Solution_210 {
    
    // tc: O(V*E)
    // sc: O(V*E)
    // took me 25 mins from reading to 1st AC
    func findOrder(_ numCourses: Int, _ prerequisites: [[Int]]) -> [Int] {
        var indegree = Counter().inDegree(prerequisites)    // what courses I need to learn
        let outdegree = Counter().outDegree(prerequisites)  // what courses need me
        
        var queue: [Int] = []   // better to use Deque, but this is not built-in DS
        var result = [Int]()
        
        // get courses which has no dependencies
        for i in 0..<numCourses {
            if indegree[i] == nil {
                queue.append(i)
            }
        }
        
        var coursesCanTake = queue.count
        
        // find circle. return []
        if coursesCanTake == 0 {
            return []
        }
        
        // BFS
        while queue.count > 0 {
            let index = queue.removeLast()
            result.append(index)
            let outs = outdegree[index]
            
            // udpate queue when course available
            if let outs = outs {
                for out in outs {
                    if indegree[out]!.count == 1 {
                        queue.append(out)
                        coursesCanTake += 1
                    } else if indegree[out]!.count > 1 {
                        indegree[out]?.remove(object: index)
                    }
                }
            }
        }
        
        if result.count == numCourses {
            return result
        } else {
            return []
        }
    }
}
